File: services/identity-service/src/test/java/br/com/deltaglobalbank/identity/features/auth/refresh/RefreshTokenTests.java
package br.com.deltaglobalbank.identity.features.auth.refresh;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.doNothing;
import static org.mockito.Mockito.lenient;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import java.time.Duration;
import java.time.Instant;
import java.util.List;
import java.util.Set;
import java.util.UUID;

import br.com.deltaglobalbank.identity.domain.tenant.Tenant;
import br.com.deltaglobalbank.identity.domain.tenant.TenantRepository;
import br.com.deltaglobalbank.identity.domain.token.InvalidRefreshTokenException;
import br.com.deltaglobalbank.identity.domain.token.RefreshToken;
import br.com.deltaglobalbank.identity.domain.token.RefreshTokenRepository;
import br.com.deltaglobalbank.identity.domain.token.RefreshTokenReuseDetectedException;
import br.com.deltaglobalbank.identity.domain.user.User;
import br.com.deltaglobalbank.identity.domain.user.UserRepository;
import br.com.deltaglobalbank.identity.domain.user.UserStatus;
import br.com.deltaglobalbank.identity.infrastructure.persistence.entities.RoleEntity;
import br.com.deltaglobalbank.identity.infrastructure.persistence.entities.TenantModuleEntity;
import br.com.deltaglobalbank.identity.infrastructure.persistence.entities.UserRoleEntity;
import br.com.deltaglobalbank.identity.infrastructure.persistence.repositories.JpaIssuedTokenAuditRepository;
import br.com.deltaglobalbank.identity.infrastructure.persistence.repositories.JpaModuleRepository;
import br.com.deltaglobalbank.identity.infrastructure.persistence.repositories.JpaRoleRepository;
import br.com.deltaglobalbank.identity.infrastructure.persistence.repositories.JpaTenantModuleRepository;
import br.com.deltaglobalbank.identity.infrastructure.persistence.repositories.JpaUserRoleRepository;
import br.com.deltaglobalbank.identity.infrastructure.security.jwt.IssuedJwt;
import br.com.deltaglobalbank.identity.infrastructure.security.jwt.JwtIssuer;
import br.com.deltaglobalbank.identity.infrastructure.security.jwt.JwtIssuerProperties;
import br.com.deltaglobalbank.identity.infrastructure.security.jwt.UserClaims;
import br.com.deltaglobalbank.identity.infrastructure.security.token.GeneratedRefreshToken;
import br.com.deltaglobalbank.identity.infrastructure.security.token.RefreshTokenGenerator;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.EnumSource;
import org.mockito.ArgumentCaptor;

class RefreshTokenTests {

    private final RefreshTokenRepository refreshTokenRepository = mock(RefreshTokenRepository.class);
    private final UserRepository userRepository = mock(UserRepository.class);
    private final TenantRepository tenantRepository = mock(TenantRepository.class);
    private final JpaUserRoleRepository userRoleRepository = mock(JpaUserRoleRepository.class);
    private final JpaRoleRepository roleRepository = mock(JpaRoleRepository.class);
    private final JpaTenantModuleRepository tenantModuleRepository = mock(JpaTenantModuleRepository.class);
    private final JpaModuleRepository moduleRepository = mock(JpaModuleRepository.class);
    private final JpaIssuedTokenAuditRepository issuedTokenAuditRepository = mock(JpaIssuedTokenAuditRepository.class);
    private final RefreshTokenGenerator refreshTokenGenerator = mock(RefreshTokenGenerator.class);
    private final JwtIssuer jwtIssuer = mock(JwtIssuer.class);
    private final JwtIssuerProperties jwtProperties = mock(JwtIssuerProperties.class);

    private RefreshTokenUseCase useCase;
    private RefreshTokenContext context;
    private UUID userId;
    private UUID tenantId;
    private Instant expiresAt;
    private UUID enabledModuleId;
    private UUID disabledModuleId;

    @BeforeEach
    void setUp() {
        useCase = new RefreshTokenUseCase(
            refreshTokenRepository, userRepository, tenantRepository, userRoleRepository, roleRepository,
            tenantModuleRepository, moduleRepository, issuedTokenAuditRepository, refreshTokenGenerator,
            jwtIssuer, jwtProperties
        );
        userId = UUID.randomUUID();
        tenantId = UUID.randomUUID();
        expiresAt = Instant.now().plusSeconds(3600);
        context = new RefreshTokenContext("192.168.0.1", "test-agent");
        lenient().when(refreshTokenGenerator.hash(any())).thenReturn("algum-hash");
        enabledModuleId = UUID.randomUUID();
        disabledModuleId = UUID.randomUUID();
    }

    @Test
    void mustThrowInvalidRefreshTokenExceptionWhenTokenIsNotFound() {
        when(refreshTokenRepository.findByTokenHash(any())).thenReturn(null);

        assertThrows(InvalidRefreshTokenException.class, () -> useCase.execute("raw-token", context));

        verify(refreshTokenRepository, times(0)).save(any());
        verify(jwtIssuer, times(0)).issueForUser(any());
    }

    @Test
    void mustRevokeAllActiveTokensAndThrowWhenReuseIsDetected() {
        RefreshToken existing = mock(RefreshToken.class);
        when(existing.isRevoked()).thenReturn(true);
        when(existing.getUserId()).thenReturn(userId);
        when(refreshTokenRepository.findByTokenHash(any())).thenReturn(existing);

        RefreshToken active1 = mock(RefreshToken.class);
        RefreshToken active2 = mock(RefreshToken.class);
        when(refreshTokenRepository.findAllActiveByUserId(userId)).thenReturn(List.of(active1, active2));
        when(refreshTokenRepository.saveAll(any())).thenReturn(List.of(active1, active2));

        assertThrows(RefreshTokenReuseDetectedException.class, () -> useCase.execute("raw-token", context));

        verify(active1, times(1)).revoke();
        verify(active2, times(1)).revoke();
        verify(refreshTokenRepository, times(1)).saveAll(any());

        verify(jwtIssuer, times(0)).issueForUser(any());
        verify(issuedTokenAuditRepository, times(0)).save(any());
    }

    @Test
    void mustThrowReuseExceptionWithoutSavingWhenNoActiveTokensExist() {
        RefreshToken existing = mock(RefreshToken.class);
        when(existing.isRevoked()).thenReturn(true);
        when(existing.getUserId()).thenReturn(userId);
        when(refreshTokenRepository.findByTokenHash(any())).thenReturn(existing);
        when(refreshTokenRepository.findAllActiveByUserId(userId)).thenReturn(List.of());

        assertThrows(RefreshTokenReuseDetectedException.class, () -> useCase.execute("raw-token", context));

        verify(refreshTokenRepository, times(0)).saveAll(any());
        verify(jwtIssuer, times(0)).issueForUser(any());
    }

    @Test
    void mustThrowInvalidRefreshTokenExceptionWhenTokenIsExpired() {
        RefreshToken existing = mock(RefreshToken.class);
        when(existing.isRevoked()).thenReturn(false);
        when(existing.isExpired()).thenReturn(true);
        when(refreshTokenRepository.findByTokenHash(any())).thenReturn(existing);

        assertThrows(InvalidRefreshTokenException.class, () -> useCase.execute("raw-token", context));

        verify(refreshTokenRepository, times(0)).save(any());
        verify(jwtIssuer, times(0)).issueForUser(any());
    }

    @Test
    void mustThrowInvalidRefreshTokenExceptionWhenUserIsNotFound() {
        RefreshToken existing = mock(RefreshToken.class);
        when(existing.isRevoked()).thenReturn(false);
        when(existing.isExpired()).thenReturn(false);
        when(existing.getUserId()).thenReturn(userId);
        when(refreshTokenRepository.findByTokenHash(any())).thenReturn(existing);
        when(userRepository.findById(any())).thenReturn(null);

        assertThrows(InvalidRefreshTokenException.class, () -> useCase.execute("raw-token", context));

        verify(refreshTokenRepository, times(0)).save(any());
        verify(jwtIssuer, times(0)).issueForUser(any());
    }

    @Test
    void mustThrowInvalidRefreshTokenExceptionWhenTenantIsNotFound() {
        RefreshToken existing = mock(RefreshToken.class);
        when(existing.isRevoked()).thenReturn(false);
        when(existing.isExpired()).thenReturn(false);
        when(existing.getUserId()).thenReturn(userId);
        when(refreshTokenRepository.findByTokenHash(any())).thenReturn(existing);

        User user = mock(User.class);
        when(user.getTenantId()).thenReturn(UUID.randomUUID());
        when(userRepository.findById(any())).thenReturn(user);

        when(tenantRepository.findById(any())).thenReturn(null);

        assertThrows(InvalidRefreshTokenException.class, () -> useCase.execute("raw-token", context));

        verify(refreshTokenRepository, times(0)).save(any());
        verify(jwtIssuer, times(0)).issueForUser(any());
    }

    @Test
    void mustThrowInvalidRefreshTokenExceptionWhenTenantIsInactive() {
        RefreshToken existing = mock(RefreshToken.class);
        when(existing.isRevoked()).thenReturn(false);
        when(existing.isExpired()).thenReturn(false);
        when(existing.getUserId()).thenReturn(userId);
        when(refreshTokenRepository.findByTokenHash(any())).thenReturn(existing);

        User user = mock(User.class);
        when(user.getTenantId()).thenReturn(UUID.randomUUID());
        when(userRepository.findById(any())).thenReturn(user);

        Tenant tenant = mock(Tenant.class);
        when(tenant.isActive()).thenReturn(false);
        when(tenantRepository.findById(any())).thenReturn(tenant);
        when(user.getId()).thenReturn(UUID.randomUUID());
        when(userRoleRepository.findAllByUserId(any())).thenReturn(List.of());

        assertThrows(InvalidRefreshTokenException.class, () -> useCase.execute("raw-token", context));

        verify(refreshTokenRepository, times(0)).save(any());
        verify(jwtIssuer, times(0)).issueForUser(any());
    }

    @ParameterizedTest
    @EnumSource(value = UserStatus.class, names = {"SUSPENDED", "LOCKED"})
    void mustThrowInvalidRefreshTokenExceptionWhenUserIsNotActive(UserStatus status) {
        RefreshToken existing = mock(RefreshToken.class);
        when(existing.isRevoked()).thenReturn(false);
        when(existing.isExpired()).thenReturn(false);
        when(existing.getUserId()).thenReturn(userId);
        when(refreshTokenRepository.findByTokenHash(any())).thenReturn(existing);

        User user = mock(User.class);
        when(user.getTenantId()).thenReturn(UUID.randomUUID());
        when(user.isActive()).thenReturn(status == UserStatus.ACTIVE);
        when(userRepository.findById(any())).thenReturn(user);

        Tenant tenant = mock(Tenant.class);
        when(tenant.isActive()).thenReturn(true);
        when(tenantRepository.findById(any())).thenReturn(tenant);
        when(user.getId()).thenReturn(UUID.randomUUID());
        when(userRoleRepository.findAllByUserId(any())).thenReturn(List.of());

        assertThrows(InvalidRefreshTokenException.class, () -> useCase.execute("raw-token", context));

        verify(refreshTokenRepository, times(0)).save(any());
        verify(jwtIssuer, times(0)).issueForUser(any());
    }

    @Test
    void mustIssueNewTokensAndRevokeCurrentTokenOnSuccessfulRefresh() {
        RefreshToken existing = mock(RefreshToken.class);
        when(existing.isRevoked()).thenReturn(false);
        when(existing.isExpired()).thenReturn(false);
        when(existing.getUserId()).thenReturn(userId);
        when(existing.getExpiresAt()).thenReturn(expiresAt);
        doNothing().when(existing).markUsed();
        doNothing().when(existing).revoke();
        when(refreshTokenRepository.findByTokenHash(any())).thenReturn(existing);

        User user = mock(User.class);
        when(user.getId()).thenReturn(userId);
        when(user.getTenantId()).thenReturn(tenantId);
        when(user.isActive()).thenReturn(true);
        when(user.mustChangePassword()).thenReturn(false);
        when(userRepository.findById(userId)).thenReturn(user);

        Tenant tenant = mock(Tenant.class);
        when(tenant.isActive()).thenReturn(true);
        when(tenantRepository.findById(tenantId)).thenReturn(tenant);

        when(tenantModuleRepository.findAllByTenantIdAndEnabled(tenantId, true)).thenReturn(List.of());
        when(userRoleRepository.findAllByUserId(userId)).thenReturn(List.of());

        IssuedJwt issuedJwt = new IssuedJwt("JWT-ACCESS-TOKEN", UUID.randomUUID(), Instant.now(), Instant.now().plusSeconds(900));
        when(jwtIssuer.issueForUser(any())).thenReturn(issuedJwt);

        GeneratedRefreshToken generated = new GeneratedRefreshToken("PLAIN-TOKEN", "HASH-INTERNO");
        when(refreshTokenGenerator.generate()).thenReturn(generated);
        when(jwtProperties.accessTokenTtl()).thenReturn(Duration.ofSeconds(900));

        ArgumentCaptor<RefreshToken> savedTokens = ArgumentCaptor.forClass(RefreshToken.class);
        when(refreshTokenRepository.save(savedTokens.capture())).thenReturn(mock(RefreshToken.class));
        when(issuedTokenAuditRepository.save(any())).thenReturn(null);

        RefreshTokenResult response = useCase.execute("raw-token", context);

        Assertions.assertAll(
            () -> assertEquals("JWT-ACCESS-TOKEN", response.accessToken()),
            () -> assertEquals("PLAIN-TOKEN", response.refreshToken()),
            () -> assertEquals("Bearer", response.tokenType()),
            () -> assertEquals(900, response.expiresIn()),
            () -> assertEquals(false, response.mustChangePassword())
        );

        RefreshToken novoToken = savedTokens.getAllValues().get(1);
        Assertions.assertAll(
            () -> assertEquals("HASH-INTERNO", novoToken.getTokenHash()),
            () -> assertEquals("192.168.0.1", novoToken.getIpAddress()),
            () -> assertEquals("test-agent", novoToken.getUserAgent())
        );

        verify(existing, times(1)).revoke();
        verify(jwtIssuer, times(1)).issueForUser(any());
        verify(issuedTokenAuditRepository, times(1)).save(any());
        verify(refreshTokenRepository, times(2)).save(any());
    }

    @Test
    void mustIncludeOnlyRolesFromEnabledModulesInJwtClaimsOnRefresh() {
        RefreshToken existing = mock(RefreshToken.class);
        when(existing.isRevoked()).thenReturn(false);
        when(existing.isExpired()).thenReturn(false);
        when(existing.getUserId()).thenReturn(userId);
        when(existing.getExpiresAt()).thenReturn(expiresAt);
        when(refreshTokenRepository.findByTokenHash(any())).thenReturn(existing);

        User user = mock(User.class);
        when(user.getId()).thenReturn(userId);
        when(user.getTenantId()).thenReturn(tenantId);
        when(user.isActive()).thenReturn(true);
        when(user.mustChangePassword()).thenReturn(false);
        when(userRepository.findById(userId)).thenReturn(user);

        Tenant tenant = mock(Tenant.class);
        when(tenant.isActive()).thenReturn(true);
        when(tenantRepository.findById(tenantId)).thenReturn(tenant);

        TenantModuleEntity enabledModule = mock(TenantModuleEntity.class);
        when(enabledModule.getModuleId()).thenReturn(enabledModuleId);
        when(tenantModuleRepository.findAllByTenantIdAndEnabled(tenantId, true)).thenReturn(List.of(enabledModule));
        when(moduleRepository.findAllById(any())).thenReturn(List.of());

        UserRoleEntity link1 = mock(UserRoleEntity.class);
        UserRoleEntity link2 = mock(UserRoleEntity.class);
        UserRoleEntity link3 = mock(UserRoleEntity.class);
        when(userRoleRepository.findAllByUserId(userId)).thenReturn(List.of(link1, link2, link3));

        RoleEntity roleEnabled = mock(RoleEntity.class);
        when(roleEnabled.getCode()).thenReturn("customers.admin");
        when(roleEnabled.getModuleId()).thenReturn(enabledModuleId);
        RoleEntity roleDisabled = mock(RoleEntity.class);
        when(roleDisabled.getCode()).thenReturn("lending.admin");
        when(roleDisabled.getModuleId()).thenReturn(disabledModuleId);
        RoleEntity roleGlobal = mock(RoleEntity.class);
        when(roleGlobal.getCode()).thenReturn("platform.admin");
        when(roleGlobal.getModuleId()).thenReturn(null);
        when(roleRepository.findAllById(any())).thenReturn(List.of(roleEnabled, roleDisabled, roleGlobal));

        ArgumentCaptor<UserClaims> claimsCaptor = ArgumentCaptor.forClass(UserClaims.class);
        IssuedJwt issuedJwt = new IssuedJwt("JWT", UUID.randomUUID(), Instant.now(), Instant.now().plusSeconds(900));
        when(jwtIssuer.issueForUser(claimsCaptor.capture())).thenReturn(issuedJwt);

        when(refreshTokenGenerator.generate()).thenReturn(new GeneratedRefreshToken("PLAIN", "HASH"));
        when(refreshTokenRepository.save(any())).thenReturn(mock(RefreshToken.class));
        when(issuedTokenAuditRepository.save(any())).thenReturn(null);
        when(jwtProperties.accessTokenTtl()).thenReturn(Duration.ofSeconds(900));

        useCase.execute("raw-token", context);

        List<String> roles = claimsCaptor.getValue().roles();
        Assertions.assertAll(
            () -> assertTrue(roles.contains("customers.admin")),
            () -> assertTrue(roles.contains("platform.admin")),
            () -> assertFalse(roles.contains("lending.admin"))
        );
    }
}

File: services/identity-service/src/test/kotlin/br/com/deltaglobalbank/identity/features/auth/refresh/RefreshTokenTests.kt
package br.com.deltaglobalbank.identity.features.auth.refresh

import br.com.deltaglobalbank.identity.domain.tenant.Tenant
import br.com.deltaglobalbank.identity.domain.tenant.TenantRepository
import br.com.deltaglobalbank.identity.domain.token.InvalidRefreshTokenException
import br.com.deltaglobalbank.identity.domain.token.RefreshToken
import br.com.deltaglobalbank.identity.domain.token.RefreshTokenRepository
import br.com.deltaglobalbank.identity.domain.token.RefreshTokenReuseDetectedException
import br.com.deltaglobalbank.identity.domain.user.User
import br.com.deltaglobalbank.identity.domain.user.UserRepository
import br.com.deltaglobalbank.identity.domain.user.UserStatus
import br.com.deltaglobalbank.identity.infrastructure.persistence.entities.RoleEntity
import br.com.deltaglobalbank.identity.infrastructure.persistence.entities.TenantModuleEntity
import br.com.deltaglobalbank.identity.infrastructure.persistence.entities.UserRoleEntity
import br.com.deltaglobalbank.identity.infrastructure.persistence.repositories.JpaIssuedTokenAuditRepository
import br.com.deltaglobalbank.identity.infrastructure.persistence.repositories.JpaModuleRepository
import br.com.deltaglobalbank.identity.infrastructure.persistence.repositories.JpaRoleRepository
import br.com.deltaglobalbank.identity.infrastructure.persistence.repositories.JpaTenantModuleRepository
import br.com.deltaglobalbank.identity.infrastructure.persistence.repositories.JpaUserRoleRepository
import br.com.deltaglobalbank.identity.infrastructure.security.jwt.IssuedJwt
import br.com.deltaglobalbank.identity.infrastructure.security.jwt.JwtIssuer
import br.com.deltaglobalbank.identity.infrastructure.security.jwt.JwtIssuerProperties
import br.com.deltaglobalbank.identity.infrastructure.security.jwt.UserClaims
import br.com.deltaglobalbank.identity.infrastructure.security.token.GeneratedRefreshToken
import br.com.deltaglobalbank.identity.infrastructure.security.token.RefreshTokenGenerator
import io.mockk.Runs
import io.mockk.every
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import io.mockk.just
import io.mockk.mockk
import io.mockk.slot
import io.mockk.verify
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertAll
import org.junit.jupiter.api.extension.ExtendWith
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.EnumSource
import java.time.Instant
import java.util.UUID

@ExtendWith(MockKExtension::class)
class RefreshTokenTests {

    @MockK
    lateinit var refreshTokenRepository: RefreshTokenRepository
    @MockK
    lateinit var userRepository: UserRepository
    @MockK
    lateinit var tenantRepository: TenantRepository
    @MockK
    lateinit var userRoleRepository: JpaUserRoleRepository
    @MockK
    lateinit var roleRepository: JpaRoleRepository
    @MockK
    lateinit var tenantModuleRepository: JpaTenantModuleRepository
    @MockK
    lateinit var moduleRepository: JpaModuleRepository
    @MockK
    lateinit var issuedTokenAuditRepository: JpaIssuedTokenAuditRepository
    @MockK
    lateinit var refreshTokenGenerator: RefreshTokenGenerator
    @MockK
    lateinit var jwtIssuer: JwtIssuer
    @MockK
    lateinit var jwtProperties: JwtIssuerProperties

    private lateinit var useCase: RefreshTokenUseCase
    private lateinit var context: RefreshTokenContext
    private lateinit var userId: UUID
    private lateinit var tenantId: UUID
    private lateinit var expiresAt: Instant
    private lateinit var enabledModuleId: UUID
    private lateinit var disabledModuleId: UUID

    @BeforeEach
    fun setUp() {
        useCase = RefreshTokenUseCase(
            refreshTokenRepository,
            userRepository,
            tenantRepository,
            userRoleRepository,
            roleRepository,
            tenantModuleRepository,
            moduleRepository,
            issuedTokenAuditRepository,
            refreshTokenGenerator,
            jwtIssuer,
            jwtProperties

        )
        userId = UUID.randomUUID()
        tenantId = UUID.randomUUID()
        expiresAt = Instant.now().plusSeconds(3600)
        context = RefreshTokenContext(ipAddress = "192.168.0.1", userAgent = "test-agent")
        every { refreshTokenGenerator.hash(any()) } returns "algum-hash"
        enabledModuleId = UUID.randomUUID()
        disabledModuleId = UUID.randomUUID()

    }

    @Test
    fun `must throw InvalidRefreshTokenException when token is not found`() {
        every { refreshTokenRepository.findByTokenHash(any()) } returns null

        assertThrows(InvalidRefreshTokenException::class.java) {
            useCase.execute("raw-token", context)
        }

        verify(exactly = 0) { refreshTokenRepository.save(any()) }
        verify(exactly = 0) { jwtIssuer.issueForUser(any()) }
    }

    @Test
    fun `must revoke all active tokens and throw when reuse is detected`() {
        val existing = mockk<RefreshToken>()
        every { existing.isRevoked() } returns true
        every { existing.userId } returns userId
        every { refreshTokenRepository.findByTokenHash(any()) } returns existing

        val active1 = mockk<RefreshToken>(relaxed = true)
        val active2 = mockk<RefreshToken>(relaxed = true)
        every { refreshTokenRepository.findAllActiveByUserId(userId) } returns listOf(active1, active2)
        every { refreshTokenRepository.saveAll(any()) } returns listOf(active1, active2)

        assertThrows(RefreshTokenReuseDetectedException::class.java) {
            useCase.execute("raw-token", context)
        }

        verify { active1.revoke() }
        verify { active2.revoke() }
        verify(exactly = 1) { refreshTokenRepository.saveAll(any()) }

        verify(exactly = 0) { jwtIssuer.issueForUser(any()) }
        verify(exactly = 0) { issuedTokenAuditRepository.save(any()) }
    }

    @Test
    fun `must throw reuse exception without saving when no active tokens exist`() {
        val existing = mockk<RefreshToken>()
        every { existing.isRevoked() } returns true
        every { existing.userId } returns userId
        every { refreshTokenRepository.findByTokenHash(any()) } returns existing
        every { refreshTokenRepository.findAllActiveByUserId(userId) } returns emptyList()

        assertThrows(RefreshTokenReuseDetectedException::class.java) {
            useCase.execute("raw-token", context)
        }

        verify(exactly = 0) { refreshTokenRepository.saveAll(any()) }
        verify(exactly = 0) { jwtIssuer.issueForUser(any()) }
    }

    @Test
    fun `must throw InvalidRefreshTokenException when token is expired`() {
        val existing = mockk<RefreshToken>()
        every { existing.isRevoked() } returns false
        every { existing.isExpired() } returns true
        every { refreshTokenRepository.findByTokenHash(any()) } returns existing

        assertThrows(InvalidRefreshTokenException::class.java) {
            useCase.execute("raw-token", context)
        }

        verify(exactly = 0) { refreshTokenRepository.save(any()) }
        verify(exactly = 0) { jwtIssuer.issueForUser(any()) }
    }

    @Test
    fun `must throw InvalidRefreshTokenException when user is not found`(){
        val existing = mockk<RefreshToken>()
        every { existing.isRevoked() } returns false
        every { existing.isExpired() } returns false
        every { existing.userId } returns userId
        every { refreshTokenRepository.findByTokenHash(any()) } returns existing
        every { userRepository.findById(any()) } returns null

        assertThrows(InvalidRefreshTokenException::class.java) {
            useCase.execute("raw-token", context)
        }

        verify(exactly = 0) { refreshTokenRepository.save(any()) }
        verify(exactly = 0) { jwtIssuer.issueForUser(any()) }
    }

    @Test
    fun `must throw InvalidRefreshTokenException when tenant is not found`() {
        val existing = mockk<RefreshToken>()
        every { existing.isRevoked() } returns false
        every { existing.isExpired() } returns false
        every { existing.userId } returns userId
        every { refreshTokenRepository.findByTokenHash(any()) } returns existing

        val user = mockk<User>()
        every { user.tenantId } returns UUID.randomUUID()
        every { userRepository.findById(any()) } returns user

        every { tenantRepository.findById(any()) } returns null

        assertThrows(InvalidRefreshTokenException::class.java) {
            useCase.execute("raw-token", context)
        }

        verify(exactly = 0) { refreshTokenRepository.save(any()) }
        verify(exactly = 0) { jwtIssuer.issueForUser(any()) }
    }

    @Test
    fun `must throw InvalidRefreshTokenException when tenant is inactive`() {
        val existing = mockk<RefreshToken>()
        every { existing.isRevoked() } returns false
        every { existing.isExpired() } returns false
        every { existing.userId } returns userId
        every { refreshTokenRepository.findByTokenHash(any()) } returns existing

        val user = mockk<User>()
        every { user.tenantId } returns UUID.randomUUID()
        every { userRepository.findById(any()) } returns user

        val tenant = mockk<Tenant>()
        every { tenant.isActive() } returns false
        every { tenantRepository.findById(any()) } returns tenant
        every { user.id } returns UUID.randomUUID()
        every { userRoleRepository.findAllByUserId(any()) } returns emptyList()

        assertThrows(InvalidRefreshTokenException::class.java) {
            useCase.execute("raw-token", context)
        }

        verify(exactly = 0) { refreshTokenRepository.save(any()) }
        verify(exactly = 0) { jwtIssuer.issueForUser(any()) }
    }

    @ParameterizedTest
    @EnumSource(value = UserStatus::class, names = ["SUSPENDED", "LOCKED"])
    fun `must throw InvalidRefreshTokenException when user is not active`(status: UserStatus) {
        val existing = mockk<RefreshToken>()
        every { existing.isRevoked() } returns false
        every { existing.isExpired() } returns false
        every { existing.userId } returns userId
        every { refreshTokenRepository.findByTokenHash(any()) } returns existing

        val user = mockk<User>()
        every { user.tenantId } returns UUID.randomUUID()
        every { user.isActive() } returns (status == UserStatus.ACTIVE)
        every { userRepository.findById(any()) } returns user

        val tenant = mockk<Tenant>()
        every { tenant.isActive() } returns true
        every { tenantRepository.findById(any()) } returns tenant
        every { user.id } returns UUID.randomUUID()
        every { userRoleRepository.findAllByUserId(any()) } returns emptyList()

        assertThrows(InvalidRefreshTokenException::class.java) {
            useCase.execute("raw-token", context)
        }

        verify(exactly = 0) { refreshTokenRepository.save(any()) }
        verify(exactly = 0) { jwtIssuer.issueForUser(any()) }
    }

    @Test
    fun `must issue new tokens and revoke current token on successful refresh`() {
        val existing = mockk<RefreshToken>()
        every { existing.isRevoked() } returns false
        every { existing.isExpired() } returns false
        every { existing.userId } returns userId
        every { existing.expiresAt } returns expiresAt
        every { existing.markUsed() } just Runs
        every { existing.revoke() } just Runs
        every { refreshTokenRepository.findByTokenHash(any()) } returns existing

        val user = mockk<User>()
        every { user.id } returns userId
        every { user.tenantId } returns tenantId
        every { user.isActive() } returns true
        every { user.mustChangePassword() } returns false
        every { userRepository.findById(userId) } returns user

        val tenant = mockk<Tenant>()
        every { tenant.isActive() } returns true
        every { tenantRepository.findById(tenantId) } returns tenant

        every { tenantModuleRepository.findAllByTenantIdAndEnabled(tenantId, true) } returns emptyList()
        every { userRoleRepository.findAllByUserId(userId) } returns emptyList()

        val issuedJwt = IssuedJwt(
            token = "JWT-ACCESS-TOKEN",
            jti = UUID.randomUUID(),
            issuedAt = Instant.now(),
            expiresAt = Instant.now().plusSeconds(900)
        )
        every { jwtIssuer.issueForUser(any()) } returns issuedJwt

        val generated = GeneratedRefreshToken(plainText = "PLAIN-TOKEN", hash = "HASH-INTERNO")
        every { refreshTokenGenerator.generate() } returns generated
        every { jwtProperties.accessTokenTtl } returns java.time.Duration.ofSeconds(900)

        val savedTokens = mutableListOf<RefreshToken>()
        every { refreshTokenRepository.save(capture(savedTokens)) } returns mockk(relaxed = true)
        every { issuedTokenAuditRepository.save(any()) } returns mockk(relaxed = true)

        val response = useCase.execute("raw-token", context)

        assertAll(
            { assertEquals("JWT-ACCESS-TOKEN", response.accessToken) },
            { assertEquals("PLAIN-TOKEN", response.refreshToken) },
            { assertEquals("Bearer", response.tokenType) },
            { assertEquals(900, response.expiresIn) },
            { assertEquals(false, response.mustChangePassword) }
        )

        val novoToken = savedTokens[1]
        assertAll(
            { assertEquals("HASH-INTERNO", novoToken.tokenHash) },
            { assertEquals("192.168.0.1", novoToken.ipAddress) },
            { assertEquals("test-agent", novoToken.userAgent) }
        )

        verify { existing.revoke() }
        verify(exactly = 1) { jwtIssuer.issueForUser(any()) }
        verify(exactly = 1) { issuedTokenAuditRepository.save(any()) }
        verify(exactly = 2) { refreshTokenRepository.save(any()) }
    }

    @Test
    fun `must include only roles from enabled modules in jwt claims on refresh`() {
        val existing = mockk<RefreshToken>(relaxed = true)
        every { existing.isRevoked() } returns false
        every { existing.isExpired() } returns false
        every { existing.userId } returns userId
        every { existing.expiresAt } returns expiresAt
        every { refreshTokenRepository.findByTokenHash(any()) } returns existing

        val user = mockk<User>()
        every { user.id } returns userId
        every { user.tenantId } returns tenantId
        every { user.isActive() } returns true
        every { user.mustChangePassword() } returns false
        every { userRepository.findById(userId) } returns user

        val tenant = mockk<Tenant>()
        every { tenant.isActive() } returns true
        every { tenantRepository.findById(tenantId) } returns tenant

        val enabledModule = mockk<TenantModuleEntity>()
        every { enabledModule.moduleId } returns enabledModuleId
        every { tenantModuleRepository.findAllByTenantIdAndEnabled(tenantId, true) } returns listOf(enabledModule)
        every { moduleRepository.findAllById(any()) } returns emptyList()

        val link1 = mockk<UserRoleEntity>(relaxed = true)
        val link2 = mockk<UserRoleEntity>(relaxed = true)
        val link3 = mockk<UserRoleEntity>(relaxed = true)
        every { userRoleRepository.findAllByUserId(userId) } returns listOf(link1, link2, link3)

        val roleEnabled = mockk<RoleEntity>()
        every { roleEnabled.code } returns ("customers.admin")
        every { roleEnabled.moduleId } returns enabledModuleId
        val roleDisabled = mockk<RoleEntity>()
        every { roleDisabled.code } returns ("lending.admin")
        every { roleDisabled.moduleId } returns disabledModuleId
        val roleGlobal = mockk<RoleEntity>()
        every { roleGlobal.code } returns ("platform.admin")
        every { roleGlobal.moduleId } returns null
        every { roleRepository.findAllById(any()) } returns listOf(roleEnabled, roleDisabled, roleGlobal)

        val claimsSlot = slot<UserClaims>()
        val issuedJwt = IssuedJwt("JWT", UUID.randomUUID(), Instant.now(), Instant.now().plusSeconds(900))
        every { jwtIssuer.issueForUser(capture(claimsSlot)) } returns issuedJwt

        every { refreshTokenGenerator.generate() } returns GeneratedRefreshToken("PLAIN", "HASH")
        every { refreshTokenRepository.save(any()) } returns mockk(relaxed = true)
        every { issuedTokenAuditRepository.save(any()) } returns mockk(relaxed = true)
        every { jwtProperties.accessTokenTtl } returns java.time.Duration.ofSeconds(900)

        useCase.execute("raw-token", context)

        val roles = claimsSlot.captured.roles
        assertAll(
            { assertTrue(roles.contains("customers.admin")) },
            { assertTrue(roles.contains("platform.admin")) },
            { assertFalse(roles.contains("lending.admin")) }
        )
    }

}